File: cpp/code/leetCode/bfs_dfs/NumberOfProvinces.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {    
public:
    void dfs(int node, vector<vector<int>>& isConnected, vector<int>& visited) {
        visited[node] = 1;
        for (int i = 0; i < isConnected[node].size(); i++) {
            if (isConnected[node][i] == 1 && visited[i] == 0) {
                dfs(i, isConnected, visited);
            }
        }
    }

    int findCircleNum(vector<vector<int>>& isConnected) {
        //노드별로 방문여부를 저장하는 vector 초기화
        vector<int> visited(isConnected.size(), 0);        
        int answer = 0;
        for (int i = 0; i < isConnected.size(); i++) {
            if (visited[i] == 0) {
                dfs(i, isConnected, visited);
                answer++;
            }
        }
        return answer;
    }
};

int main() {
    //vector<vector<int>> isConnected = {
    //    {1, 0, 0},
    //    {0, 1, 0},
    //    {0, 0, 1}
    //};
    //[[1,0,0,1],[0,1,1,0],[0,1,1,1],[1,0,1,1]]
    vector<vector<int>> isConnected = {
        {1,0,0,0,0,0,0,0,0,1,0,0,0,0,0},
        {0,1,0,1,0,0,0,0,0,0,0,0,0,1,0},
        {0,0,1,0,0,0,0,0,0,0,0,0,0,0,0},
        {0,1,0,1,0,0,0,1,0,0,0,1,0,0,0},
        {0,0,0,0,1,0,0,0,0,0,0,0,1,0,0},
        {0,0,0,0,0,1,0,0,0,0,0,0,0,0,0},
        {0,0,0,0,0,0,1,0,0,0,0,0,0,0,0},
        {0,0,0,1,0,0,0,1,1,0,0,0,0,0,0},
        {0,0,0,0,0,0,0,1,1,0,0,0,0,0,0},
        {1,0,0,0,0,0,0,0,0,1,0,0,0,0,0},
        {0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
        {0,0,0,1,0,0,0,0,0,0,0,1,0,0,0},
        {0,0,0,0,1,0,0,0,0,0,0,0,1,0,0},
        {0,1,0,0,0,0,0,0,0,0,0,0,0,1,0},
        {0,0,0,0,0,0,0,0,0,0,0,0,0,0,1}
    };
    
    Solution s;
    
    cout << s.findCircleNum(isConnected);

	return 0;
}

File: rust/src/leet_code/bfs_dfs/NumberOfProvinces.rs
struct Solution;

impl Solution {
    fn dfs(node : usize, is_connected: &Vec<Vec<i32>>, visited: &mut Vec<i32>) {
        visited[node] = 1;
        for i in 0 .. is_connected[node].len(){
            //print!("{}", is_connected[node][i]);
            if is_connected[node][i] == 1 && visited[i] == 0 {
                Self::dfs(i, &is_connected, visited);
            }
        }
    }
    pub fn find_circle_num(is_connected: Vec<Vec<i32>>) -> i32 {
        let mut visited : Vec<i32> = vec![0; is_connected.len()];
        let mut answer : i32 = 0;
        for i in 0 .. is_connected.len(){
            if visited[i] == 0 {
                Self::dfs(i, &is_connected, &mut visited);
                answer += 1;
            }
            //println!();
        }
        answer
    }
}
pub fn run(){
    // let mut is_connected: Vec<Vec<i32>> = Vec::new();
    // is_connected.push(vec![1, 1, 0]);
    // is_connected.push(vec![1, 1, 0]);
    // is_connected.push(vec![0, 0, 1]);

    let is_connected : Vec<Vec<i32>> = vec![
        vec![1,0,0,0,0,0,0,0,0,1,0,0,0,0,0],
        vec![0,1,0,1,0,0,0,0,0,0,0,0,0,1,0],
        vec![0,0,1,0,0,0,0,0,0,0,0,0,0,0,0],
        vec![0,1,0,1,0,0,0,1,0,0,0,1,0,0,0],
        vec![0,0,0,0,1,0,0,0,0,0,0,0,1,0,0],
        vec![0,0,0,0,0,1,0,0,0,0,0,0,0,0,0],
        vec![0,0,0,0,0,0,1,0,0,0,0,0,0,0,0],
        vec![0,0,0,1,0,0,0,1,1,0,0,0,0,0,0],
        vec![0,0,0,0,0,0,0,1,1,0,0,0,0,0,0],
        vec![1,0,0,0,0,0,0,0,0,1,0,0,0,0,0],
        vec![0,0,0,0,0,0,0,0,0,0,1,0,0,0,0],
        vec![0,0,0,1,0,0,0,0,0,0,0,1,0,0,0],
        vec![0,0,0,0,1,0,0,0,0,0,0,0,1,0,0],
        vec![0,1,0,0,0,0,0,0,0,0,0,0,0,1,0],
        vec![0,0,0,0,0,0,0,0,0,0,0,0,0,0,1]
    ];

    println!("{}", Solution::find_circle_num(is_connected));
}